h/services/bulk_executor/_actions.py

```diff
@@ -0,0 +1,156 @@
+        """
+        raise NotImplementedError()
+
+    @staticmethod
+    def _check_upsert_queries(batch, expected_keys):
+        """Check that the query matches expectations and the attributes.
+
+        :param batch: A collection of command objects
+        :param expected_keys: The keys which must be in the query and match
+                              the attributes with the same name.
+        """
+
+        for command in batch:
+            query = command.body.query
+
+            # This is technically overkill as the schema should make sure we
+            # can't receive queries we aren't expecting
+            if set(query.keys()) != set(expected_keys):
+                raise UnsupportedOperationError(
+                    f"Upserting by query fields '{query.keys()}' is not supported"
+                )
+
+            # Checking that the values are the same is a bit more important, as
+            # this happens post schema, and could therefore be wrong
+            for key, expected in query.items():
+                if command.body.attributes[key] != expected:
+                    raise UnsupportedOperationError(
+                        "Upserting different values to the query is not supported. "
+                        f"Different value found in key '{key}'"
+                    )
+
+    def _execute_statement(self, stmt):
+        result = self.db.execute(stmt)
+
+        # Let SQL Alchemy know that something has changed, otherwise it will
+        # never commit the transaction we are working on and it will get rolled
+        # back
+        mark_changed(self.db)
+
+        return result
+
+
+class UserUpsertAction(DBAction):
+    """Perform a bulk user upsert.
+
+    :raises ConflictingDataError: If two users attempt to use the same identity
+    """
+
+    def execute(self, batch, **_):
+        # Check that we can actually process this batch
+        self._check_upsert_queries(batch, expected_keys=["authority", "username"])
+
+        # Split users and their embedded identity lists
+        users, identities = [], []
+
+        for command in batch:
+            attributes = deepcopy(command.body.attributes)
+
+            identities.append(attributes.pop("identities"))
+            users.append(attributes)
+
+        # Upsert the data
+        user_rows = self._upsert_user_table(users)
+        self._upsert_identities(identities, user_ids=[row[0] for row in user_rows])
+
+        # Report back
+        return [
+            Report(id_, public_id=User(authority=authority, _username=username).userid)
+            for id_, authority, username in user_rows
+        ]
+
+    def _upsert_user_table(self, users):
+        stmt = self._upsert_statement(
+            User,
+            users,
+            index=[
+                # The text wrapper here prevents SQLAlchemy from quoting the
+                # string, and we need it to be verbatim
+                text("lower(replace(username, '.'::text, ''::text)), authority")
```

Comment:
I don't know, we need what this serialises as in SQL. We could try calling `text(_normalise_username())` and see if that does what we expect?

Reply:
I've tried this a few different ways and I can't get anything to work. Basically we shouldn't be putting functions in our indexes if we want anything like this to be pleasant, as it prevents us from inferring the index by column names (which is the work around to the work around of not being able to specify the index by name, because it isn't a "constraint" even though internally to Postgres it is...)

The advice I've seen is to have an additional normalised field derived from the main field with the function, and then make the compound index from that field without the function. Then we could infer the index by specifying the two fields by name.

Reply:
Fair enough